File: src/services/embeddings.ts
import { createHash } from "node:crypto";
import { HfInference } from "@huggingface/inference";
import { env } from "../config/env.js";

function meanPool(tokenVectors: number[][]) {
  const dim = tokenVectors[0]?.length ?? 0;
  const out = new Array<number>(dim).fill(0);
  for (const tv of tokenVectors) {
    for (let i = 0; i < dim; i++) out[i] += tv[i] ?? 0;
  }
  const n = tokenVectors.length || 1;
  for (let i = 0; i < dim; i++) out[i] /= n;
  return out;
}

function l2Normalize(v: number[]) {
  let sum = 0;
  for (const x of v) sum += x * x;
  const denom = Math.sqrt(sum) || 1;
  return v.map((x) => x / denom);
}

function stableTokenize(text: string) {
  return text
    .toLowerCase()
    .replace(/[`*_>#()[\]{}|~]/g, " ")
    .replace(/[^\p{L}\p{N}\s]+/gu, " ")
    .split(/\s+/g)
    .map((t) => t.trim())
    .filter((t) => t.length >= 2);
}

function tokenToIndex(token: string, dim: number) {
  // Fast, stable hash -> index
  const h = createHash("sha256").update(token).digest();
  const n = h.readUInt32LE(0);
  return n % dim;
}

function tokenToSign(token: string) {
  const h = createHash("sha256").update(`sign:${token}`).digest();
  return (h[0]! & 1) === 0 ? 1 : -1;
}

async function embedTexts(texts: string[]) {
  const dim = env.EMBEDDINGS_DIM;
  return texts.map((text) => {
    const v = new Array<number>(dim).fill(0);
    const tokens = stableTokenize(text);
    for (const token of tokens) {
      const idx = tokenToIndex(token, dim);
      v[idx] += tokenToSign(token);
    }
    return l2Normalize(v);
  });
}

export type Embeddings = {
  embedDocuments(texts: string[]): Promise<number[][]>;
  embedQuery(text: string): Promise<number[]>;
};

async function embedTextsHf(texts: string[]) {
  if (!env.HF_TOKEN) {
    throw new Error("Missing HF_TOKEN (required when EMBEDDINGS_PROVIDER=hf).");
  }
  const hf = new HfInference(env.HF_TOKEN);
  const vectors: number[][] = [];

  // Be gentle to free-tier limits.
  const batchSize = 8;
  for (let i = 0; i < texts.length; i += batchSize) {
    const batch = texts.slice(i, i + batchSize);
    for (const t of batch) {
      const res: unknown = await hf.featureExtraction({
        model: env.HF_EMBEDDINGS_MODEL,
        inputs: t
      });

      // HF feature-extraction may return:
      // - number[] (already pooled)
      // - number[][] (token vectors)
      const v =
        Array.isArray(res) && typeof (res as any)[0] === "number"
          ? (res as number[])
          : Array.isArray(res) && Array.isArray((res as any)[0])
            ? meanPool(res as number[][])
            : null;

      if (!v) {
        throw new Error(`Unexpected HF featureExtraction response shape for model ${env.HF_EMBEDDINGS_MODEL}`);
      }
      vectors.push(l2Normalize(v));
    }
  }

  return vectors;
}

export function getEmbeddings(): Embeddings {
  if (env.EMBEDDINGS_PROVIDER === "hf") {
    return {
      embedDocuments: async (texts) => embedTextsHf(texts),
      embedQuery: async (text) => (await embedTextsHf([text]))[0]!
    };
  }
  return {
    embedDocuments: async (texts) => embedTexts(texts),
    embedQuery: async (text) => (await embedTexts([text]))[0]!
  };
}

File: app/embeddings.py
import hashlib
import os
import re
from typing import List

from huggingface_hub import InferenceClient


def _l2_normalize(v: List[float]) -> List[float]:
    s = sum(x * x for x in v) ** 0.5
    if s == 0:
        return v
    return [x / s for x in v]


def _stable_tokenize(text: str) -> List[str]:
    text = text.lower()
    text = re.sub(r"[`*_>#()\\[\\]{}|~]", " ", text)
    text = re.sub(r"[^\w\s]+", " ", text, flags=re.UNICODE)
    tokens = [t.strip() for t in re.split(r"\s+", text) if t.strip()]
    return [t for t in tokens if len(t) >= 2]


def _sha256_bytes(s: str) -> bytes:
    return hashlib.sha256(s.encode("utf-8")).digest()


def _token_to_index(token: str, dim: int) -> int:
    h = _sha256_bytes(token)
    n = int.from_bytes(h[:4], "little", signed=False)
    return n % dim


def _token_to_sign(token: str) -> int:
    h = _sha256_bytes("sign:" + token)
    return 1 if (h[0] & 1) == 0 else -1


def embed_offline(texts: List[str], dim: int) -> List[List[float]]:
    out: List[List[float]] = []
    for text in texts:
        v = [0.0] * dim
        for token in _stable_tokenize(text):
            v[_token_to_index(token, dim)] += float(_token_to_sign(token))
        out.append(_l2_normalize(v))
    return out


def _mean_pool(tokens: List[List[float]]) -> List[float]:
    if not tokens:
        return []
    dim = len(tokens[0])
    acc = [0.0] * dim
    for tv in tokens:
        for i in range(dim):
            acc[i] += float(tv[i])
    n = float(len(tokens))
    return [x / n for x in acc]


def embed_hf(texts: List[str], hf_token: str, model: str) -> List[List[float]]:
    if not hf_token:
        raise ValueError("Missing HF_TOKEN (required when EMBEDDINGS_PROVIDER=hf).")
    client = InferenceClient(provider="hf-inference", api_key=hf_token)
    out: List[List[float]] = []
    for t in texts:
        res = client.feature_extraction(t, model=model)
        if res and isinstance(res[0], (float, int)):
            v = [float(x) for x in res]  # type: ignore[arg-type]
        else:
            v = _mean_pool(res)  # type: ignore[arg-type]
        out.append(_l2_normalize(v))
    return out


def embed_texts(texts: List[str]) -> List[List[float]]:
    provider = os.getenv("EMBEDDINGS_PROVIDER", "offline").strip().lower()
    if provider == "hf":
        return embed_hf(
            texts=texts,
            hf_token=os.getenv("HF_TOKEN", "").strip(),
            model=os.getenv("HF_EMBEDDINGS_MODEL", "BAAI/bge-large-en").strip(),
        )
    dim = int(os.getenv("EMBEDDINGS_DIM", "512"))
    return embed_offline(texts=texts, dim=dim)

